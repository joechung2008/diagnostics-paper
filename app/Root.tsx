import React, { startTransition, useCallback, useState } from "react";
import { StyleSheet, View } from "react-native";
import { SegmentedButtons } from "react-native-paper";
import useSWR from "swr";
import Header from "../components/Header";
import TabPanels from "../components/TabPanels";
import { Environment, EnvironmentType } from "../lib/environment";
import type { ExtensionInfo, KeyedNavLink } from "../lib/types";
import { fetchDiagnostics, isExtensionInfo } from "../lib/utils";

const tabButtons = [
  { value: "extensions", label: "Extensions" },
  { value: "build", label: "Build Info" },
  { value: "server", label: "Server Info" },
];

const Root: React.FC = () => {
  const [environment, setEnvironment] = useState<EnvironmentType>(
    Environment.Public
  );

  const {
    data: diagnostics,
    error,
    isLoading,
  } = useSWR(environment, fetchDiagnostics);

  const [extension, setExtension] = useState<ExtensionInfo>();

  const [selectedTab, setSelectedTab] = useState<string>("extensions");

  const [viewMode, setViewMode] = useState<"list" | "details">("list");

  const handleEnvironmentChange = useCallback((value: string) => {
    startTransition(() => {
      setEnvironment(value as EnvironmentType);
      setExtension(undefined);
      setViewMode("list");
    });
  }, []);

  const handleExtensionSelect = useCallback((extension: ExtensionInfo) => {
    setExtension(extension);
    setSelectedTab("extensions");
    setViewMode("details");
  }, []);

  const handleLinkClick = useCallback(
    (item: KeyedNavLink) => {
      if (item) {
        const $extension = diagnostics?.extensions[item.key];
        if (isExtensionInfo($extension)) {
          handleExtensionSelect($extension);
        }
      }
    },
    [diagnostics?.extensions, handleExtensionSelect]
  );

  const handleTabClick = useCallback(
    (value: string) => {
      // If clicking the already-selected Extensions tab, toggle back to list view
      if (value === "extensions" && selectedTab === "extensions") {
        setViewMode("list");
      }
      setSelectedTab(value);
    },
    [selectedTab]
  );

  return (
    <View style={styles.container}>
      <Header
        environment={environment}
        diagnostics={diagnostics}
        onEnvironmentChange={handleEnvironmentChange}
        onExtensionSelect={handleExtensionSelect}
      />

      <SegmentedButtons
        buttons={tabButtons}
        style={styles.segmentedButtons}
        value={selectedTab}
        onValueChange={handleTabClick}
      />

      <TabPanels
        isLoading={isLoading}
        error={error}
        selectedTab={selectedTab}
        diagnostics={diagnostics}
        viewMode={viewMode}
        extension={extension}
        onLinkClick={handleLinkClick}
        onViewModeChange={setViewMode}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  segmentedButtons: {
    margin: 10,
  },
});

export default Root;
