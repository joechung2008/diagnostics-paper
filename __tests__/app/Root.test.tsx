import React from "react";
import { render } from "@testing-library/react-native";
import useSWR from "swr";
import Root from "../../app/Root";

// Mock useSWR
jest.mock("swr", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("Root", () => {
  beforeEach(() => {
    (useSWR as jest.MockedFunction<typeof useSWR>).mockReturnValue({
      data: {
        extensions: {},
        build: {},
        server: {},
      },
      error: null,
      isLoading: false,
      mutate: jest.fn(),
      isValidating: false,
    });
  });

  it("renders without crashing", () => {
    const { toJSON } = render(<Root />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders Header component", () => {
    const { getByTestId } = render(<Root />);
    // Assuming Header has a testID, but since it's mocked, perhaps check for text or something
    // For now, just check it renders
    expect(getByTestId).toBeDefined(); // Placeholder
  });

  // Add more tests as needed
});
