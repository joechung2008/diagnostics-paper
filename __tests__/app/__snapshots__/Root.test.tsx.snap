// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root renders without crashing 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      {
        "backgroundColor": "#FFFBFE",
      },
    ]
  }
>
  Appbar.Header: [object Object],[object Object],false,[object Object]
  SegmentedButtons: Extensions, Build Info, Server Info
  <RCTScrollView
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          {
            "padding": 10,
          }
        }
      >
        <RCTScrollView
          accessibilityLabel="Extensions"
          style={
            {
              "backgroundColor": "#FFFBFE",
              "flexGrow": 0,
              "flexShrink": 1,
              "minWidth": 0,
              "padding": 10,
            }
          }
        >
          <View />
        </RCTScrollView>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
